teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+      type: string
+  - name: flags
+    description: ""
+    schema:
+      type: array
+  - name: sdist
+    description: ""
+    schema:
+      type: object
+      properties:
+        sha: {
+          type: string
+        }
+        path: {
+          type: string
+        }
+```
+
+We could have:
+
+```yaml
+params:
+  - name: someURL
+    description: ""
+    type: string
+  - name: flags
+    description: ""
+    type: array
+  - name: sdist
+    description: ""
+    type: dict # use dict intead of object
+    keys: # use keys instead of properties
+      sha: {
+        description: "" # add description section for each key
+      }
+      path: { }
+  optional: [ ] # add "optional" and have required by default
+```
+
+But the big differences are cosmetic (dict vs object, keys vs properties), and with the current proposal we could
+reasonably add our own `description` field into the existing property dictionaries (and `description` is already
+supported by [Open API schema!](#why-json-schema).
+
+## Alternative #3: Create our own syntax just for dictionaries
+
+We could make our own syntax, specific to dictionaries, however if we later add more JSON Schema support we'd need to
+revisit this. For example:
+
+```yaml
+ results:
+   - name: sdist
+     type: dict
+     keys: [ 'sha', 'path' ]
+```
+
+This is clean and clear when we're talking about a dictionary of just strings, but what if we want to allow nesting,
+e.g. array or dict values?
+
+### More alternatives
+
+1. Add support for nested types right away
+1. Add complete JSON Schema support
+1. Use something other than JSON Schema (or our own syntax proposed in Alternative #1) for expressing the
+   object/dictionary structure It feels at this point like it's better to adopt a syntax that already handles these
+   cases. (See [why JSON Schema](#why-json-schema))
+1. Reduce the scope, e.g.:
+    * Don't add support for grabbing individual items in a dictionary
+        * We need this eventually for the future to be useful
+    * Don't let dictionaries declare their keys (e.g. avoid bringing in JSON Schema)
+        * Without a programmatic way to know what keys a dictionary needs/provides, we actually take a step backward in
+          clarity, i.e. a list of strings has more information than a dictionary where you don't know the keys
+
+## Upgrade & Migration Strategy
+
+[As mentioned in TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) this update will
+be completely backwards compatible.
+
+## Potential future work
+
+Once Task and Pipeline authors are able to define object schemas for Tasks and Params, it would be very useful to:
+
+1. Allow definition reuse instead of having to copy and paste them around
+2. If we ship a known set of these with the Tekton Pipelines controller, they could define interfaces that tools such
+   as [Tekton Chains](https://github.com/tektoncd/chains/blob/main/docs/config.md) could rely on and build around.
```

Comment:
+100

I didn't realise this was already part of the TEP :) 
I agree it can be future work, but I think it would be an important feature to have

Reply:
💯  - @wlynch and his team are planning to create a TEP to dive into this in more detail